amr-wind/equation_systems/icns/source_terms/DragForcing.cpp

```diff
@@ -0,0 +1,153 @@
+    const amrex::Real gpu_drag = m_drag;
+    const amrex::Real gpu_spongeStrength = m_spongeStrength;
+    const amrex::Real gpu_spongeDensity = m_spongeDensity;
+    const amrex::Real gpu_startX = (m_spongeDistanceX > 0)
+                                       ? prob_hi[0] - m_spongeDistanceX
+                                       : prob_lo[0] - m_spongeDistanceX;
+    const amrex::Real gpu_startY = (m_spongeDistanceY > 0)
+                                       ? prob_hi[1] - m_spongeDistanceY
+                                       : prob_lo[1] - m_spongeDistanceY;
+    const amrex::Real gpu_spongeX = (m_spongeDistanceX > 0) ? 1 : 0;
+    const amrex::Real gpu_spongeY = (m_spongeDistanceY > 0) ? 1 : 0;
+    // Copy Data
+    const auto* local_gpu_vel_ht = gpu_vel_ht.data();
+    const auto* local_gpu_vel_vals = gpu_vel_vals.data();
+    const unsigned vsize = gpu_vel_ht.size();
+    amrex::ParallelFor(bx, [=] AMREX_GPU_DEVICE(int i, int j, int k) noexcept {
+        const amrex::Real x1 = prob_lo[0] + (i + 0.5) * dx[0];
+        const amrex::Real x2 = prob_lo[1] + (j + 0.5) * dx[1];
+        const amrex::Real x3 = prob_lo[2] + (k + 0.5) * dx[2];
+        amrex::Real xdamping = 0;
+        amrex::Real ydamping = 0;
+        amrex::Real xi = (gpu_spongeX == 1)
+                             ? (x1 - gpu_startX) / (prob_hi[0] - gpu_startX)
+                             : (gpu_startX - x1) / (gpu_startX - prob_lo[0]);
+        xi = std::max(xi, 0.0);
+        xdamping = gpu_spongeStrength * xi * xi;
+        amrex::Real yi = (gpu_spongeY == 1)
+                             ? (x2 - gpu_startY) / (prob_hi[1] - gpu_startY)
+                             : (gpu_startY - x2) / (gpu_startY - prob_lo[1]);
+        yi = std::max(yi, 0.0);
+        ydamping = gpu_spongeStrength * yi * yi;
+        amrex::Real Cd = gpu_drag / dx[0];
+        amrex::Real gpu_spongeVelX = 0.0;
+        amrex::Real gpu_spongeVelY = 0.0;
+        amrex::Real gpu_spongeVelZ = 0.0;
+        amrex::Real residual = 1000;
+        amrex::Real height_error = 0.0;
+        for (unsigned ii = 0; ii < vsize; ++ii) {
+            height_error = std::abs(x3 - local_gpu_vel_ht[ii]);
+            if (height_error < residual) {
+                residual = height_error;
+                const unsigned ix = 3 * ii;
+                const unsigned iy = 3 * ii + 1;
+                const unsigned iz = 3 * ii + 2;
+                gpu_spongeVelX = local_gpu_vel_vals[ix];
+                gpu_spongeVelY = local_gpu_vel_vals[iy];
+                gpu_spongeVelZ = local_gpu_vel_vals[iz];
+            }
+        }
+        // Terrain Drag
+        amrex::Real Dxz = 0.0;
+        amrex::Real Dyz = 0.0;
+        const amrex::Real ux1 = vel(i, j, k, 0);
+        const amrex::Real uy1 = vel(i, j, k, 1);
+        const amrex::Real uz1 = vel(i, j, k, 2);
+        const amrex::Real m = std::sqrt(ux1 * ux1 + uy1 * uy1 + uz1 * uz1);
+        if (drag(i, j, k) == 1) {
```

Comment:
Similar to Michael's comments, is `drag` an integer/bool field? Also, perhaps call this something more descriptive than `drag` to not confuse this with `m_drag`?